### PR TITLE
Feature/225 - submit button for facet filter bar

### LINKED ISSUE #225: Improve Facet Options Filter Bar
Need to improve the usability of the facet options text field for filtering options -- right now it loses focus on every update.

### PR DESCRIPTION
**Pull request by doug-patterson**
Fixes #225 

* this works as-is, but there are a number of questions
  * should we abstract `FacetOptionsFilter` more and move it somewhere else?
    * didn't seem worth it to me, but I may be an outlier on that
  * should we include the old implementation and maybe default to it?
  * if so, where should we turn the new opts filtering on?
      * could pass something for it from the theme to the exampleTypes constructor
  * what should the text on the submit button be? should you be able to pass it in?

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Your PR has lint errors. Please fix these and commit them.
      
<details>
  <summary>Facet.js</summary>
  <p><a href="https://github.com/smartprocure/contexture-react/blob/feature/225/src/exampleTypes/Facet.js" target="_blank">src/exampleTypes/Facet.js</a></p>
  
  <strong>'e' is defined but never used.</strong>
  <code>Line 71: <Button {...buttonEnabled ? {} : { disabled: true }} onClick={e => buttonEnabled && tree.mutate(node.path, { optionsFilter: val })}></code>
</details>
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>The changelog has not been updated. Please update the changelog.</td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>The version was not updated. Please update the version.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Could not run `duti:fix` command successfully
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7a880a9b88b5c80c2bb07aeecf68381a480b7b6f
</p>

File: src/exampleTypes/Facet.js [hunk 1]
```diff
@@ -53,13 +53,37 @@ let SelectAll = observer(({ node, tree, Checkbox }) => {
     </label>
   )
 })
+
+let FacetOptionsFilter = ({ tree, node, TextInput, Button }) => {
+  let [val, setVal] = useState(node.optionsFilter)
+  let [buttonEnabled, setButtonEnabled] = useState(!!_.size(node.optionsFilter))
```
Comment (daedalus28) [inline src/exampleTypes/Facet.js]: buttonEnabled can just be a "computed" as `let buttonEnabled = val === node.optionsFilter`

File: src/exampleTypes/Facet.js [hunk 2]
```diff
@@ -53,13 +53,37 @@ let SelectAll = observer(({ node, tree, Checkbox }) => {
...
+  return (
+    <Observer>
+      {() => <Flex style={{ justifyContent: 'space-between' }}>
+        <TextInput
+          value={val}
+          onChange={e => {
+            setVal(e.target.value)
+            setButtonEnabled(_.size(e.target.value || val))
```
Comment (daedalus28) [inline src/exampleTypes/Facet.js]: if you do the above change, you won't need setButtonEnabled here at all

File: src/exampleTypes/Facet.js [hunk 3]
```diff
@@ -53,13 +53,37 @@ let SelectAll = observer(({ node, tree, Checkbox }) => {
...
+          }}
+          placeholder="Find..."
+        />
+        <Button {...buttonEnabled ? {} : { disabled: true }} onClick={e => buttonEnabled && tree.mutate(node.path, { optionsFilter: val })}>
```
Comment (daedalus28) [inline src/exampleTypes/Facet.js]: you could just do `disabled={!buttonEnabled}`

File: src/exampleTypes/Facet.js (hunk 1 above)
Comment (doug-patterson) [reply src/exampleTypes/Facet.js]: @daedalus28 good point

**Comment by doug-patterson** [comment]
@daedalus28 updated

**Comment by daedalus28** [comment]
@doug-patterson there's a lint error and version conflicts, but otherwise, I think this is good to go and not WIP anymore, right?

**Comment by daedalus28** [comment]
Maybe just add a couple of screenshots to get product to sign off before it gets merged

**Comment by doug-patterson** [comment]
Screenshot 2019-05-14 11 16 07
Screenshot 2019-05-14 11 16 20
Screenshot 2019-05-14 11 16 28

**Comment by doug-patterson** [comment]
@daedalus28 updated

**Comment by doug-patterson** [comment]
@daedalus28 updated

**Comment by daedalus28** [comment]
OK - @doug-patterson finally got some UX feedback here.

Let's hide the button instead of disabled when not needed and use the gv-button group thing to make them look like they're attached instead of just touching

**Comment by geosp** [comment]
Screen Shot 2019-05-30 at 4 29 18 PM
Screen Shot 2019-05-30 at 4 29 42 PM

**Comment by geosp** [comment]
Screen Shot 2019-05-30 at 4 32 43 PM

**Pull request merged**